examples/data-objects/inventory-app/src/legacySharedTreeInventoryList.ts

```diff
@@ -0,0 +1,136 @@
+			undefined,
+			LegacySharedTree.getFactory().type,
+		) as LegacySharedTree;
+
+		// Initialize the inventory with two parts at zero quantity
+		// REV: Is ChangeNode appropriate here? Or should I use some other type (BuildNode?)?
+		const inventoryNode: ChangeNode = {
+			identifier: legacySharedTree.generateNodeId(),
+			// REV: This is based on the Definition casting from Bubblebench, it doesn't really seem
+			// to matter what is used here though.  Is there a better way to omit the definition here?
+			definition: "array" as Definition,
+			traits: {
+				nut: [
+					{
+						identifier: legacySharedTree.generateNodeId(),
+						definition: "scalar" as Definition,
+						traits: {},
+						payload: 0,
+					},
+				],
+				bolt: [
+					{
+						identifier: legacySharedTree.generateNodeId(),
+						definition: "scalar" as Definition,
+						traits: {},
+						payload: 0,
+					},
+				],
+			},
+		};
+		legacySharedTree.applyEdit(
+			Change.insertTree(
+				inventoryNode,
+				StablePlace.atStartOf({
+					parent: legacySharedTree.currentView.root,
+					label: "parts" as TraitLabel,
+				}),
+			),
+		);
+
+		this.root.set(legacySharedTreeKey, legacySharedTree.handle);
+	}
+
+	protected async hasInitialized() {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		this._tree = await this.root
+			.get<IFluidHandle<LegacySharedTree>>(legacySharedTreeKey)!
+			.get();
+
+		// REV: I see the tsdoc for EditCommitted recommends using a Checkout instead but that seems
+		// like overkill (plus I don't see how it relates exactly).  Is there a better way to just observe
+		// that the scalars are changing and raise "inventoryChanged"?
+		this._tree.on(SharedTreeEvent.EditCommitted, () => {
+			this.emit("inventoryChanged");
+		});
+	}
+
+	// REV: Lots of casts and non-null assertions below - is there a more type-confident way to do this?
+	public getParts() {
+		const parts: IPart[] = [];
+		// REV: Seems strange that this.tree.currentView.rootNode is private.
+		const rootNode = this.tree.currentView.getViewNode(this.tree.currentView.root);
+		const partsNode = this.tree.currentView.getViewNode(
+			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+			rootNode.traits.get("parts" as TraitLabel)![0],
+		);
+		// REV: This seems to iterate in reverse order?  Not sure why it swaps the entries as compared to how
+		// it was created.
+		for (const [partLabel, [partQuantityNodeId]] of partsNode.traits) {
+			const partQuantityNode = this.tree.currentView.getViewNode(partQuantityNodeId);
+			const quantity = partQuantityNode.payload as number;
+			// REV: It would probably be preferable to have a durable IPart that raises individual
+			// "quantityChanged" events rather than eventing/refreshing on a whole-tree basis.  Is
+			// there a good way to listen for tree edits under a specific node?  Is the partQuantityNode
+			// reference durable across tree changes?
```

Comment:
It's possible to do this relatively efficiently with legacy shared-tree, but the machinery has to be implemented on consumer side. I believe new shared-tree's eventing API does allow scoped subscription.

In terms of durability, node ids are stable across all edits, so the callbacks you're attaching are fine to use across edits and should do what you want, you just won't have all of the right nodes (or too many, if some have been deleted since the last call to `getParts()`). The view nodes are still valid to call APIs on, but probably aren't what you want (they do not get updated in response to changes--the underlying data structures powering the view are copy-on-write-like)

Reply:
You probably don't want to use the edit log. The intended way to figure out which nodes are changed (and query against the tree for what happened) is via `TreeView.delta`. I know it's an extra step, but you'll want to make a checkout for this and subscribe to the 'viewChange' event to do this.

Some sample code for the same pet project which does this is [here](https://github.com/Abe27342/belle-puzzles/blob/f04567ab9e1d977f4112d9530421db4a0932d1ce/packages/bot/src/sync/viewChangeHandler.ts#L40).

I can also point you to a couple internal legacy shared-tree usages over teams if you want to look at those (but the diff logic is a lot more sprawling)

Reply:
> Re: durability: If I'm holding refs to two nodes A and B, and A updates - are both refs now "stale" or is my B ref still good?

What types are you holding onto exactly and how is one 'updating'?

Reply:
Gotcha - I'll look into the checkout stuff then.

In this scenario if I have a node for nut quantity and a node for bolt quantity and am holding references to both - and then I update bolt quantity - I think the bolt quantity node is stale.  But I'm unclear if the nut quantity node reference is still good or is also stale and needs to be refreshed from an updated view.

Reply:
The underlying data structure each of those nodes is part of is copy-on-write, so you can always query any node you have (more details on that structure [here](https://github.com/qwertie/btree-typescript#features)). You're right that the bolt quantity will be stale, but typically you'd refresh that by getting a checkout viewChange event, calling `new.delta(old)`, seeing that the id for the bolt node is 'changed' (if you're representing the quantity as a payload on that node; if you do other edits maybe you'd see a delete/add for nested node if you had a boxed number, etc.), and realize you need to react in your application accordingly.

If you don't see a portion of the tree as added/deleted/changed while calling delta, and none of its ancestors have changed, it's safe to assume they are still valid and up-to-date content within the tree.

Reply:
I think I have this mostly working now - is there an equivalent in new SharedTree to find the changed nodes/delta (currently I'm just using "afterBatch")?

Reply:
I see Craig's PR uses a "subtreeChanging" that sounds promising, I'll take a look at that.